llumnix/queue/zmq_client.py

```diff
@@ -21,105 +22,160 @@
-        # ZMQ_SOCKET_LIMIT (http://api.zeromq.org/4-2:zmq-ctx-get)
-        socket_limit = self.context.get(zmq.constants.SOCKET_LIMIT)
-        if socket_limit < RPC_SOCKET_LIMIT_CUTOFF:
-            raise ValueError(
-                f"Found zmq.constants.SOCKET_LIMIT={socket_limit}, which caps "
-                "the number of concurrent requests Llumnix can process.")
-
-        # We only have 1 ipc connection that uses unix sockets, so
-        # safe to set MAX_SOCKETS to the zmq SOCKET_LIMIT (i.e. will
-        # not run into ulimit issues)
-        self.context.set(zmq.constants.MAX_SOCKETS, socket_limit)
+        self.context = zmq.asyncio.Context(8)
+        self.socket_pool_factory: ZmqSocketPoolFactory = ZmqSocketPoolFactory(
+            context=self.context, pool_size=SOCKET_POOL_MAXSIZE
+        )
+        self.zmq_timeout_ms: int = RPC_GET_DATA_TIMEOUT_MS
+        self._conn_lock: asyncio.Lock = Lock()
 
     # This function is not called explicitly.
     def close(self):
+        self.socket_pool_factory.close_all_pools()
         self.context.destroy()
 
-    @contextmanager
-    def to_socket(self, rpc_path):
-        # Raise a sensible error if the client was already closed.
-        # This can happen if a server shutdown is triggered but some coroutines
-        # are still running requests.
-        # There should not be a race condition with this check because we don't
-        # yield to the event loop between here and opening the socket.
-        if self.context.closed:
-            raise RPCClientClosedError("The ZMQ client has already shut down")
-
-        # Note that we use DEALER to enable asynchronous communication
-        # to enable streaming.
-        socket = self.context.socket(zmq.constants.DEALER)
-        socket.set_hwm(RPC_ZMQ_HWM)
-        try:
-            socket.connect(rpc_path)
-            yield socket
-        finally:
-            socket.close(linger=0)
-
     async def _send_one_way_rpc_request(
             self,
             request: RPC_REQUEST_TYPE,
-            rpc_path: str,
+            ip: str,
+            port: int,
             error_message: str):
         async def do_rpc_call(socket: zmq.asyncio.Socket,
                               request: RPC_REQUEST_TYPE):
 
             await socket.send_multipart([cloudpickle.dumps(request)])
 
-            if await socket.poll(timeout=self._data_timeout) == 0:
+            if await socket.poll(timeout=self.zmq_timeout_ms) == 0:
                 raise TimeoutError("Server didn't reply within "
-                                   f"{self._data_timeout} ms")
+                                    f"{self.zmq_timeout_ms} ms")
 
             return cloudpickle.loads(await socket.recv())
 
-        with self.to_socket(rpc_path) as socket:
-            response = await do_rpc_call(socket, request)
+        socket_pool = await self.socket_pool_factory.get_pool(ip, port)
+        socket = await socket_pool.get_socket()
```

Comment:
i have a main concern of your pr, that is we only need to make io operation to be coroutine in order to achieve high throughput, but it seems that there are some non-io functions being coroutine in your pr. For example, the queue storing socket is also a asyncio queue, is it necessary?